File: LC/[0759] Employee Free Time.py
###############
# 20231226
###############

"""
# Definition for an Interval.
class Interval:
    def __init__(self, start: int = None, end: int = None):
        self.start = start
        self.end = end
"""


class Solution:
    def employeeFreeTime(self, schedule: '[[Interval]]') -> '[Interval]':
        sweep_line = []
        for sch in schedule:
            for interval in sch:
                sweep_line.append((interval.start, 1))
                sweep_line.append((interval.end, -1))

        sweep_line.sort(key=lambda x: (x[0], -x[1]))

        st = -1
        cnt = 0
        res = []
        for tt, inc in sweep_line:
            cnt += inc

            if cnt == 0 and inc == -1:
                st = tt

            if cnt == 1 and inc == 1 and st != -1:
                res.append(Interval(st, tt))
                st = -1

        return res


####################
"""
# Definition for an Interval.
class Interval:
    def __init__(self, start: int = None, end: int = None):
        self.start = start
        self.end = end
"""


class Solution:
    def employeeFreeTime(self, schedule: '[[Interval]]') -> '[Interval]':
        sweep_line = []
        for sch in schedule:
            for interval in sch:
                sweep_line.append((interval.start, 1))
                sweep_line.append((interval.end, -1))

        # increase first before decrease
        sweep_line.sort(key=lambda x: (x[0], -x[1]))

        res = []
        cnt = 0
        st, ed = -1, -1
        for t, v in sweep_line:
            cnt += v

            if cnt == 0 and v == -1:
                st = t
            elif cnt == 1 and v == 1:
                ed = t
                if st != -1:
                    res.append(Interval(st, ed))
        return res

File: LC/[0759] Employee Free Time.java
/*
// Definition for an Interval.
class Interval {
    public int start;
    public int end;

    public Interval() {}

    public Interval(int _start, int _end) {
        start = _start;
        end = _end;
    }
};
*/


class Solution {

    // Sort by time asc; for ties, put +1 before -1 (descending by inc),
    // matching: if (a.first != b.first) a.first < b.first else a.second > b.second
    public List<Interval> employeeFreeTime(List<List<Interval>> schedule) {
        List<int[]> events = new ArrayList<>(); // [time, inc]
        for (List<Interval> emp : schedule) {
            for (Interval r : emp) {
                events.add(new int[]{r.start, +1});
                events.add(new int[]{r.end, -1});
            }
        }

        events.sort((a, b) -> {
            if (a[0] != b[0]) return Integer.compare(a[0], b[0]);
            return Integer.compare(b[1], a[1]); // inc desc: +1 before -1
        });

        int count = 0;
        Integer start = null; // start of a free interval (null means none yet)
        List<Interval> ret = new ArrayList<>();

        for (int[] e : events) {
            count += e[1];
            if (e[1] == -1 && count == 0) {
                // just transitioned to no active work intervals → free time starts
                start = e[0];
            } else if (e[1] == +1 && count == 1 && start != null) {
                // just transitioned from free (0) to busy (1) → free time ends
                int end = e[0];
                ret.add(new Interval(start, end));
                start = null;
            }
        }
        return ret;
    }
}